### DISCUSSION TITLE
Use of assignment expressions :=

### DISCUSSION THREAD

**Discussion by cdce8p** [home-assistant/architecture#617] [General] [answered]
For the next pylint version I will add a new optional checker (https://github.com/PyCQA/pylint/pull/4876) to detect if assignment expressions can be used instead of an assignment followed by an `if`. I was wondering if that is something we want to enable in the core repo.

There would be about 1500 places where it could be used.
An example commit with changes for a few select files can be found here: https://github.com/cdce8p/ha-core/commit/eda27e1fad2c0f8b75887f7d0fb3d44f9d2ffe2e

--
Some assignment expressions are already used, the difference would be that with the check it would become much more widespread.

--
**Update**
As of pylint 2.11, the `consider-using-assignment-expr` check is part of the pylint `CodeStyle` extension and would only need to be activated. I've opened https://github.com/home-assistant/core/pull/56383 for it.

Before moving forward and making assignment expressions mandatory, with some exceptions (see below), I would still like to get some more feedback. https://github.com/home-assistant/core/pull/56394 added the proposed `:=` to a few files - I think it helps to see the potential impact for this change.

**Exception**: `consider-using-assignment-expr` is only emitted if the resulting line is `<= 72` characters. That limit is configurable (in `pyproject.toml`) but from my testing anything over that will become too difficult to read.

**Comment by ludeeus** [comment]
Personally, I prefer the use of assignment expressions, since it reduces lines of code.
But it should only be done if the value is actually accessed after, which is not the case several places in <https://github.com/cdce8p/ha-core/commit/eda27e1fad2c0f8b75887f7d0fb3d44f9d2ffe2e>

```python
if not (safe_mode := runtime_config.safe_mode):
```
Should be:
```python
if not runtime_config.safe_mode:
```
***
```python
        if (flow := self._progress.get(flow_id)) is None:
            raise UnknownFlow
```
Should be:
```python
        if self._progress.get(flow_id) is None:
            raise UnknownFlow
```
or :
```python
        if not self._progress.get(flow_id):
            raise UnknownFlow
```
***
```python
            if (dep_itg := integration_cache.get(domain)) is None:
                continue
``` 
Should be:
```python
            if integration_cache.get(domain) is None:
                continue
``` 
or:
```python
            if integration_cache.get(domain) is None:
                continue
```

**Reply by cdce8p** [reply]
@ludeeus In general I agree with you. I did take another look at all places you mentioned: In every case the variable is used later on, just not inside the `if` block itself.

**Reply by ludeeus** [reply]
~~How? We raise/continue?~~

I see what you mean 👍

**Reply by cdce8p** [reply]
The assignment happens even if the `if` check itself isn't successful.

**Comment by janiversen** [comment]
My preference is clearly assignment, because it is less code and a lot easier to read. The use of the := operator is something you can easily overlook and thus dangerous.

**Reply by cdce8p** [reply]
I disagree. Once you know what to look for, assignment expressions can be quite useful. It's just initially that they might seem a bit overwhelming. Please take a look at https://github.com/home-assistant/core/pull/56394 for some examples.

**Reply by chemelli74** [reply]
I vote to use walrus as well.

Simone

**Reply by Kane610** [reply]
I'm with the walrus!

**Reply by janiversen** [reply]
I did not vote for making assignment expressions illegal, I just believe it is not wise to enforce their use by adding the check to pylint.

I have my preference as you have yours, do we really need to restrict that.

**Reply by cdce8p** [reply]
You can still use `pylint: disable`. However I believe that I've already written the check in a way to only add a warning for the most obvious examples. E.g. there is a `max-line-length` setting just for suggestions. I explained it here https://github.com/home-assistant/core/pull/56364#discussion_r711580127, but basically if using `:=` would result in a line longer than `72` chars, no warning will be emitted. Again please look at https://github.com/home-assistant/core/pull/56394 to see for yourself.

The main advantage I see with the pylint check is that it will automatically detect places where `:=` could be useful. One thing less to think about during a code review.

**Reply by thecode** [reply]
Coming from C background I did not like walrus at first, but Python is not C and already using walrus since it is already in many places in the code, So I promote using walrus, but I don't think pylint should force using it.

**Reply by Jc2k** [reply]
I think walrus simplifies code and makes it shorter and easier to read. We should use it of course.

For me the question is: does the check automate something that would be flagged up in review. Would I call out code that could be written better with walrus? Heck yeah I would. As a reviewer personal preference would not come into it, if walrus would make a function clearer and more concise I'd down vote in review and ask for a walrus version. So a check to automate this seems reasonable.

I think it's entirely reasonable to encourage walrus as part of the HA code style, and entirely reasonable to use automation to encourage people to follow that code style.

Personal preference is always at odds with a projects code style guide. But a project the size of HA benefits from leaning more towards consistency than it does towards personal preference. You benefit from following your own personal style. Everyone else benefits from a common style. Everyone else is a much bigger number.

(Tl;dr - +1 for check)

**Comment by bdraco** [comment]
I like the idea of making this a pylint check for the reasons outlined in https://www.python.org/dev/peps/pep-0572/#the-importance-of-real-code

**Comment by raman325** [comment]
I agree with this change based on all of the other comments (in addition to the OP 🙂 )

**Comment by balloob** [comment, marked as answer]
let's do it.

**Comment by thecode** [comment]
Let's do it, but we do need to think of a way to run `pylint` only on changed files in dev container pre-commit . We have today the `-> None` return type and this will also added which are not commonly used in other python projects so would be nice if pre-commit will validate it. I know there is a problem running `pylint` due to requirements missing, but for changed files requirement should be installed.

**Reply by balloob** [reply]
It's hacktoberfest, 1500 places? No problem 🤣

**Reply by scop** [reply]
> we do need to think of a way to run `pylint` only on changed files in dev container pre-commit

The usual working mode of pre-commit in dev setups where one commits is to run it on changed files only. Is devcontainer somehow different in this respect?

**Reply by thecode** [reply]
> > we do need to think of a way to run `pylint` only on changed files in dev container pre-commit
> 
> The usual working mode of pre-commit in dev setups where one commits is to run it on changed files only. Is devcontainer somehow different in this respect?

`pylint` is currently not a part of HA dev container pre-commit, in addition if the you run pylint not from the root directory our special adjustments are not in. Which means `pylint` failures are only found after a PR is submitted unless the user manually run `pylint` from root directory before pre-commit.

**Reply by scop** [reply]
Ah yes, forgot about that. Haven't looked at all into what specialties HA has there, but at least "vanilla" pylint can be made run with usual pre-commit setups, basically in similar manner as mypy.

**Reply by Jc2k** [reply]
I think this is off-topic for this issue. If you have ideas on how to improve the dev container setup we should probably do that in another ticket. I don't think we need arch issues for the dev container, so maybe we can go straight to PR's too 🥳 .

(I personally have a custom pre-commit script that runs everything (including tests and pylint) just for the integration I am working on, and then CI runs everything anyway in case that somehow doesnt get run).

**Comment by janiversen** [comment]
Core/.pre-commit-config.yaml suggest that we never (not in dev-container, nor in core) run pylint as part of the pre-commit.

I was actually told  a while ago, that the reason for this was that pylint is too slow.

I run pylint (and mypy) manually on my branches, and our “pylint” specialities are not difficult to add to devcontainer (mostly configuration and some plugins).
